src/Scrutor/ServiceCollectionExtensions.Composite.cs

```diff
@@ -0,0 +1,41 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+// ReSharper disable once CheckNamespace
+namespace Microsoft.Extensions.DependencyInjection
+{
+    public static partial class ServiceCollectionExtensions
+    {
+        /// <summary>
+        /// Registers <typeparamref name="TConcrete"/> as a composite that wraps all
+        /// existing registrations of <typeparamref name="TInterface"/>.
+        /// </summary>
+        /// <typeparam name="TInterface"></typeparam>
+        /// <typeparam name="TConcrete"></typeparam>
+        /// <param name="services"></param>
+        public static void AddComposite<TInterface, TConcrete>(this IServiceCollection services)
+          where TInterface : class
+          where TConcrete : class, TInterface
+        {
+            var wrappedDescriptors = services.Where(s => s.ServiceType == typeof(TInterface)).ToList();
+            foreach (var descriptor in wrappedDescriptors)
+                services.Remove(descriptor);
+
+            var objectFactory = ActivatorUtilities.CreateFactory(
+              typeof(TConcrete),
+              new[] { typeof(IEnumerable<TInterface>) });
+
+            var maxWrappedServiceLifetime = wrappedDescriptors
+                .Select(d => d.Lifetime)
+                .DefaultIfEmpty(ServiceLifetime.Scoped)
+                .Max();
```

Comment:
I'm not so sure about this. Would it be better to actually have explicit lifetimes, like `AddCompositeTransient`, `AddCompositeScoped` and `AddCompositeSingleton`?

Reply:
Actually, I'm not sure this will fly at all. There can be valid reasons for having mixed lifetimes in a collection of services. This will make all services a singleton if there's just a single singleton in the collection.

Reply:
Yeah, looking again I think this was based on a false assumption.  I've replaced it with explicit specification of lifetime and 3 explicit methods.

Reply:
Yeah, agreed.  I've replaced that behaviour in the the latest commit

Reply:
No, you still remove all the existing descriptors, which could have mixed lifetimes, and replace them with a single descriptor.

Reply:
Sorry, I misunderstood your point before.  You're right: I'm not sure that it's possible to avoid this behaviour.

Ideally if you have a transient composite wrapping 1 singleton and 1 transient service then when you create multiple composites you would expect only 1 singleton to ever be created.

I could "solve" this by only composing services in the same scope but that doesn't feel right and probably isn't particularly useful.

I can think of some creative ways around this but nothing simple.  Do you think it's worth looking into a potentially-convoluted solution?